lib/util/identifier.js

```diff
@@ -7,10 +7,26 @@ const looksLikeAbsolutePath = (maybeAbsolutePath) => {
 
 const normalizePathSeparator = (p) => p.replace(/\\/g, "/");
 
+/* A map from context dir strings to maps from identifier names to relative paths */
+const relativePaths = new Map();
+
 exports.makePathsRelative = (context, identifier) => {
-	return identifier
-		.split(/([|! ])/)
-		.map(str => looksLikeAbsolutePath(str) ?
-			normalizePathSeparator(path.relative(context, str)) : str)
-		.join("");
+	if(!relativePaths.has(context)) {
+		relativePaths.set(context, new Map());
+	}
+
+	const contextCache = relativePaths.get(context);
+
+	if(contextCache.has(identifier)) {
+		return contextCache.get(identifier);
+	} else {
+		var relativePath = identifier
```

Comment:
Might be nice to move this to a small helper identifierToRelativePath now that we're memoizing.